sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/AvoidExcessiveClassCoupling.cs

```diff
@@ -65,150 +67,186 @@ public sealed class AvoidExcessiveClassCoupling : ParameterLoadingDiagnosticAnal
-                        .Union(allFieldsCoupling)
-                        .Union(allPropertiesCoupling)
-                        .Union(allMethodsCoupling);
-                    var classCouplingCount = FilterAndCountCoupling(totalClassCoupling);
-
-                    if (classCouplingCount > Threshold)
+                    if (dependentTypes.Count > Threshold)
                     {
-                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
-                            classCouplingCount, Threshold));
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, typeDeclaration.Identifier.GetLocation(),
+                            typeDeclaration.Keyword.ValueText, dependentTypes.Count, Threshold));
                     }
                 },
-                SyntaxKind.ClassDeclaration);
+                SyntaxKind.ClassDeclaration,
+                SyntaxKind.StructDeclaration,
+                SyntaxKind.InterfaceDeclaration);
         }
 
-        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(FieldDeclarationSyntax field,
-            SemanticModel model)
+        private static bool IsCountableType(INamedTypeSymbol namedType) =>
+            namedType.TypeKind != TypeKind.Enum &&
+            !namedType.IsAny(ignoredTypes);
+
+        private static IEnumerable<INamedTypeSymbol> ExpandGenericTypes(INamedTypeSymbol namedType)
         {
-            return field.Declaration.Variables
-                .SelectMany(variable => ExtractTypesFromVariableDeclarator(variable, model));
+            var originalDefinition = new[] { namedType.OriginalDefinition };
+            if (!namedType.IsGenericType)
+            {
+                return originalDefinition;
+            }
+            return namedType.IsUnboundGenericType
+                ? originalDefinition.Union(namedType.TypeParameters.SelectMany(GetConstraintTypes))
+                : originalDefinition.Union(namedType.TypeArguments.OfType<INamedTypeSymbol>().SelectMany(ExpandGenericTypes));
         }
 
-        private static IEnumerable<ITypeSymbol> CollectCoupledClasses(BasePropertyDeclarationSyntax property,
-            SemanticModel model)
+        private static IEnumerable<INamedTypeSymbol> GetConstraintTypes(ITypeParameterSymbol typeParameter) =>
+            typeParameter.ConstraintTypes.OfType<INamedTypeSymbol>().SelectMany(ExpandGenericTypes);
+
+        private class TypeDependencyCollector : CSharpSyntaxWalker
         {
-            var propertyReturnType = new[] { model.GetSymbolInfo(property.Type).Symbol as ITypeSymbol };
+            private readonly SemanticModel model;
+            private readonly TypeDeclarationSyntax originalTypeDeclaration;
 
-            var indexerProperty = property as IndexerDeclarationSyntax;
-            if (indexerProperty?.ExpressionBody != null)
+            public IList<ITypeSymbol> DependentTypes { get; } = new List<ITypeSymbol>();
```

Comment:
Could this be a set rather than a list?

Reply:
it could probably save some iterations, I guess